### PR TITLE
settings

### PR DESCRIPTION
**Pull request by dyfvicture**

### CODE REVIEW

File: src/components/settings/components/Relay.js [hunk 1]
```diff
@@ -44,30 +45,29 @@ const RealySettingForm = ({
...
                         </div>
                         <div className="col mr10">
-                          <Input size="large" value="27.0.0.01" />
+                          <Input size="large" value={item.value} disabled/>
                         </div>
                         <div className="col-auto">
-                          { index >0 &&
-                            <a href="" onClick={gotoEdit} className="">Edit</a>
+                          { item.custom &&
+                            <a href="javascript:void(0)" onClick={gotoEdit.bind(this, item.id)} className="">Edit</a>
```
Comment (hzbdesigner) [inline src/components/settings/components/Relay.js]: use e.preventDefault() to fix this window default actions

File: src/components/settings/components/Preference.js [hunk 2]
```diff
@@ -2,14 +2,13 @@ import React from 'react';
...
 const Perference = ({
-    form
+    settings, form
   }) => {
-  // const {preference} = settings
+  const {preference} = settings
   function handleChange(type, value) {
-    console.log(type+":"+value);
+    console.log(type+" changed to:"+value);
+    settings.preferenceChange({...preference, [type]: value})
```
Comment (hzbdesigner) [inline src/components/settings/components/Preference.js]: no need to put other properties in preference

File: src/components/settings/components/Relay.js [hunk 3]
```diff
@@ -8,16 +8,17 @@ import RelayEdit from './RelayEdit'
...
+    modals.showModal({id:'settings/relay/edit', relayId:relayId})
   }
   const gotoAdd = ()=>{
     modals.showModal({id:'settings/relay/add'})
   }
-
-  function handleChange(type, value) {
-    console.log(type+":"+value);
+  function handleChange(e) {
+    settings.relayChange({...relay, selected:e.target.value})
```
Comment (hzbdesigner) [inline src/components/settings/components/Relay.js]: remove ...relay

File: src/components/settings/components/RelayEdit.js [hunk 4]
```diff
@@ -1,49 +1,79 @@
 import React from 'react';
 import { Form,InputNumber,Button,Icon,Modal,Input,Radio,Select,Checkbox,Slider,Card} from 'antd';
 import {languagesArray, timezoneArray} from '../../../common/config/data'
+import {url} from '../../../common/utils/regx'
 
 const RelayEditForm = ({
-    form
+    form, settings, modal, modals
```
Comment (hzbdesigner) [inline src/components/settings/components/RelayEdit.js]: remove modals

File: src/components/settings/components/RelayEdit.js [hunk 5]
```diff
@@ -1,49 +1,79 @@
...
   function handleChange(type, value) {
     console.log(type+":"+value);
   }
   function handleSubmit() {
     form.validateFields((err,values) => {
-      console.log('values',values);
       if(!err){
-        // TODO
+        settings.editRelay({id: modal.relayId, name: values.name, url: values.url})
+        modals.hideModal({id:'settings/relay/edit'})
```
Comment (hzbdesigner) [inline src/components/settings/components/RelayEdit.js]: repalce modals by modal

File: src/components/settings/components/RelayAdd.js [hunk 6]
```diff
@@ -22,21 +24,42 @@ const AddRelayForm = ({
   function resetForm(){
     form.resetFields()
   }
+  function validateRelayName(value) {
+    if(!value) return false;
+    return !relay.nodes.find(item=>item.name === value)
+  }
+  function validateUrl(value) {
```
Comment (hzbdesigner) [inline src/components/settings/components/RelayAdd.js]: use rules not custom validator

File: src/components/settings/components/Trading.js [hunk 7]
```diff
@@ -1,13 +1,39 @@
...
 const TradingSettingForm = ({
     settings,form
   }) => {
-
+  const {trading} = settings
   function handleChange(type, value) {
-    console.log(type+":"+value);
+    console.log(type+" changed to:"+value);
+    if ("contractVersion" === type){
+      const addresses = configs.contracts.filter(item => item.version === value)
```
Comment (hzbdesigner) [inline src/components/settings/components/Trading.js]: use find not filter

File: src/components/settings/components/Trading.js [hunk 8]
```diff
@@ -1,13 +1,39 @@
...
+      if(addresses.length > 0) {
+        settings.tradingChange({...trading, contract:{version:value, address:addresses[0].address}})
```
Comment (hzbdesigner) [inline src/components/settings/components/Trading.js]: remove ..trading

File: src/components/settings/components/Preference.js (hunk 2 above)
Comment (dyfvicture) [reply src/components/settings/components/Preference.js]: fixed

File: src/components/settings/components/Relay.js (hunk 3 above)
Comment (dyfvicture) [reply src/components/settings/components/Relay.js]: fixed

File: src/components/settings/components/Relay.js (hunk 1 above)
Comment (dyfvicture) [reply src/components/settings/components/Relay.js]: fixed

File: src/components/settings/components/RelayAdd.js (hunk 6 above)
Comment (dyfvicture) [reply src/components/settings/components/RelayAdd.js]: use type:'url' instead

File: src/components/settings/components/RelayEdit.js (hunk 4 above)
Comment (dyfvicture) [reply src/components/settings/components/RelayEdit.js]: fixed

File: src/components/settings/components/RelayEdit.js (hunk 5 above)
Comment (dyfvicture) [reply src/components/settings/components/RelayEdit.js]: fixed

File: src/components/settings/components/Trading.js [hunk 9]
```diff
@@ -1,13 +1,41 @@
...
 const TradingSettingForm = ({
     settings,form
   }) => {
-
-  function handleChange(type, value) {
-    console.log(type+":"+value);
+  const {trading} = settings
+  function handleChange(type, e) {
+    if ("contractVersion" === type){
+      const addresses = configs.contracts.filter(item => item.version === e)
```
Comment (hzbdesigner) [inline src/components/settings/components/Trading.js]: use arr.find instead of arr.filter

**Review by hzbdesigner — APPROVED** [review]

**Pull request merged**